src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.self_attn_layer_norm(x)
+        residual = x
+        x, _ = self.encoder_attn(
+            query=x,
+            key=encoder_hidden_states,
+            key_padding_mask=encoder_attn_mask,
+            layer_state=layer_state,  # mutates layer state
+        )
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.encoder_attn_layer_norm(x)
+
+        # Fully Connected
+        residual = x
+        x = self.activation_fn(self.fc1(x))
+        x = F.dropout(x, p=self.activation_dropout, training=self.training)
+        x = self.fc2(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.final_layer_norm(x)
+
+        return (
+            x,
+            self_attn_weights,
+            layer_state,
+        )  # just self_attn weights for now, following t5, layer_state = cache for decoding
+
+def ngram_attention_bias(length, num_skip):
+        bias_result = []
+        for n_skip in range(num_skip):
+            bias_n_skip = []
+            for i in range(length):
+                bias_this = [float('-inf')] * (2 * length)
+                bias_this[length + i] = 0
+                first_k = i - n_skip
+                first_k = first_k if first_k > 0 else 0
+                for j in range(first_k + 1):
+                    bias_this[j] = 0
+                bias_n_skip.append(bias_this)
+            bias_result.append(bias_n_skip)
+        return torch.from_numpy(np.array(bias_result, dtype=np.float32))
+
+class ProphetNetDecoder(nn.Module):
+    """
+    N-stream decoder. One main stream, self.ngram predicting streams.
+    Next self.ngram tokens are predicted.
+
+    N-stream decoder consisting of *config.decoder_layers* layers. Each layer
+    is a :class:`ProphetNetDecoderLayer`.
+    Args:
+        config: ProphetNetConfig
+        embed_tokens (torch.nn.Embedding): output embedding
+    """
+    def __init__(self, config: ProphetNetConfig, embed_tokens: nn.Embedding):
+        super().__init__()
+        self.ngram = config.ngram
+        self.num_buckets = config.num_buckets
+        self.relative_max_distance = config.relative_max_distance
+
+        # for hugging face
+        self.output_hidden_states = config.output_hidden_states
+
+        self.dropout = config.dropout
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_target_positions = config.max_position_embeddings
+        self.embed_scale = None
+        self.embed_tokens = embed_tokens
+        embed_dim = config.d_model
+        self.embed_positions = LearnedPositionalEmbedding(config.max_position_embeddings + 2 + self.padding_idx,
+                                                          embed_dim, self.padding_idx)
+        self.ngram_input_embed = nn.Embedding(self.ngram, embed_dim, None)
+        self.layers = nn.ModuleList([])
```

Comment:
can do everything in one line IMO -> no need for `self.layers = nn.ModuleList([])`